src/helpers/clipboard/clipboard_figure_state.ts

```diff
@@ -84,12 +84,20 @@ export class ClipboardFigureState implements ClipboardState {
    */
   paste(target: Zone[]) {
     const sheetId = this.getters.getActiveSheetId();
+    const size = { height: this.copiedFigure.height, width: this.copiedFigure.width };
+    const viewportRect = this.getters.getMainViewportRect();
+    const colDimensions = this.getters.getColDimensions(sheetId, target[0].left);
+    const rowDimensions = this.getters.getRowDimensions(sheetId, target[0].top);
     const position = {
-      x: this.getters.getColDimensions(sheetId, target[0].left).start,
-      y: this.getters.getRowDimensions(sheetId, target[0].top).start,
+      x:
+        viewportRect.width - colDimensions.start < size.width
```

Comment:
pretty sure you lack a condition to prevent x and y to be negative.  Why not just computing the value `viewportRect.width - (size.width + colDimensions.start)` and if it's negative, subtract if from colDimensions.start ? (resp the sale thing for `y`)